pkg/routine/redis_lock.go

```diff
@@ -0,0 +1,101 @@
+// Copyright © 2020 by PACE Telematics GmbH. All rights reserved.
+// Created at 2020/02/27 by Marius Neugebauer
+
+package routine
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/bsm/redislock"
+	"github.com/go-redis/redis"
+	redisbackend "github.com/pace/bricks/backend/redis"
+)
+
+func routineThatKeepsRunningOneInstance(name string, routine func(context.Context)) func(context.Context) {
+	return func(ctx context.Context) {
+		locker := redislock.New(getDefaultRedisClient())
+		var tryAgainIn time.Duration // zero on first run
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-time.After(tryAgainIn):
+			}
+			lockCtx := obtainLock(ctx, locker, "routine:lock:"+name, cfg.RedisLockTTL)
+			if lockCtx != nil {
+				routine(lockCtx)
+			}
+			tryAgainIn = cfg.RedisLockTTL / 2
+		}
+	}
+}
+
+var (
+	initRedisOnce sync.Once
+	redisClient   *redis.Client
+)
+
+func getDefaultRedisClient() *redis.Client {
+	initRedisOnce.Do(func() { redisClient = redisbackend.Client() })
+	return redisClient
+}
+
+// Try to obtain a lock. Return a sub-context of ctx that is canceled once the
+// lock is lost or ctx is done.
+func obtainLock(ctx context.Context, locker *redislock.Client, key string, ttl time.Duration) context.Context {
+	// obtain lock
+	lock, err := locker.Obtain(key, ttl, nil)
+	if err == redislock.ErrNotObtained {
+		return nil
+	} else if err != nil {
+		panic(err)
```

Comment:
This will crash an entire process (and potentially pod) without notifying anyone about it. It Is not obvious from the outside that a panic can occur this deep in RunNamed()